환경별 프로파일 설정

tlqkrus012345 opened the pull request:
test, local, dev 환경으로 프로파일을 나눴습니다.
추후에 운영 환경까지 고려한다면 prod 환경도 필요하다 생각했지만 현재는 운영 배포 계획이 없기 때문에 우선은 개발 및 테스트에 필요한 세 가지 환경만으로 구성했습니다.

tlqkrus012345 committed:
chore: 환경별 프로파일 설정

- 환경에 따라 다른 프로파일을 설정하여 DB 데이터 보호

Resolves: #6

tlqkrus012345 committed:
build: mysql 라이브러리 추가

Resolves: #6

tlqkrus012345 committed:
test: 테스트 프로파일 분리

- 테스트 환경에 맞는 프로파일 설정

Resolves: #6

tlqkrus012345 committed:
Merge branch 'feature/6' of https://github.com/f-lab-edu/pay-together into feature/6

f-lab-martin commented on src/test/java/com/paytogether/PaytogetherApplicationTests.java:
```diff
@@ -2,7 +2,10 @@
+import org.springframework.context.annotation.Profile;
+import org.springframework.test.context.ActiveProfiles;
 
+@ActiveProfiles("test")
```
이 코드를 사용하지 않는 방법은 없을까요? 
테스트 패키지 안에 있는 코드들은 모두 테스트 profile을 쓸 것으로 가정할 수 있을 것 같은데요.
통합테스트마다 해당 코드를 작성해야하면 비효율적이지 않을까요?

tlqkrus012345 replied:
gradle.build 파일에서 junitplatform에 한정해서 test 프로퍼티를 적용하게 변경했습니다

f-lab-martin commented on src/main/resources/application-local.yml:
```diff
@@ -0,0 +1,30 @@
+      on-profile: local
+
+  datasource:
+    url: jdbc:h2:mem:pay_together
```
로컬 환경에서도 mysql db에 연결해서 dev환경에서의 동작을 미리 테스트해보면 어떨까요?

tlqkrus012345 replied:
로컬 환경에도 mysql db를 적용했습니다!

f-lab-martin commented on src/test/java/com/paytogether/member/service/SessionLoginServiceTest.java:
```diff
@@ -15,8 +15,10 @@
 
 @DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+@ActiveProfiles("test")
 @SpringBootTest
```
각 애노테이션의 의미를 알고 계실까요?

tlqkrus012345 replied:
DirtyContext 어노테이션 의미를 이해하지 못하고 사용했습니다. 
테스트를 실행할 때 같은 값(DB uniqe key)를 사용해서 생기는 오류를 새로운 애플리케이션 컨텍스트를 생성하여 H2 DB를 초기화 했습니다. 따라서 해당 어노테이션을 제거하고 @BeforeEach 어노테이션을 사용하여 테스트 실행 전 기존 데이터를 삭제하는 기능을 추가했습니다.

f-lab-martin commented:
로컬에서 테스트하기 위해 redis를 특정 포트로 띄워야 합니다. 이 내용을 Readme.md에도 작성해서 가이드해보면 어떨까요?
혹은 docker-compose를 사용해서 infra 세팅을 편리하게 도와주면 어떨까요?

tlqkrus012345 committed:
build: 테스트 실행 시 'test'

Resolves: #6

tlqkrus012345 committed:
test: 어노테이션 제거

- @Distinct 어노테이션 -> @BeforeEach
- @ActiveProfile -> gradle에서 동작하게

Resolves: #6

tlqkrus012345 committed:
chore: local datasource 변경

- h2 -> mysql

Resolves: #6

tlqkrus012345 committed:
infra: redis, mysql docker-compose 작성

- 로컬에서 테스트 일관성을 위해 redis, mysql docker-compose 파일 작성

Resolves: #6

tlqkrus012345 committed:
fix: h2 설정 제거

- application.yml 파일에서 h2 관련 설정 제거로 인하여 Spring Security가 해당 빈을 못찾는 에러 발생하여 제거

Resolves: #6
